### ISSUE TITLE
Alignment requirements of cuda::std::complex

### ISSUE THREAD

**Issue by gonzalobg** [NVIDIA/libcudacxx#151]
The following two `static_assert`s compile without issues:

```c++
#include <cuda/std/complex>
static_assert(alignof(cuda::std::complex<double>) == 8);
static_assert(alignof(cuda::std::complex<float>) == 4);
```

I'd expected them to be 16 and 8 to match the `double2` and `float2` types.

**Comment by dkolsen-pgi** [comment]
I'd expect them to be 8 and 4 to match the `std::complex<double>` and `std::complex<float>` types.

There are competing requirements here.  It may very well be that the higher alignment is the better choice, but it is not unambiguously the right choice.

**Comment by gonzalobg** [comment]
What are the requirements for `cuda::std::complex`?

**Comment by jrhemstad** [comment]
> What are the requirements for `cuda::std::complex`?

The requirements for anything in `cuda::std` are exactly equivalent (when possible) to its counterpart in `std::`.

**Comment by cliffburdick** [comment]
To add to a separate discussion, this becomes an issue when using `cuda::std::complex` to access global memory within a warp. The unaligned accesses cause a 50% performance drop due to the uncoalesced access.

**Comment by brycelelbach** [comment]
I think we should break from what the standard requires here. The only case where it hits us is if someone is casting unaligned `T[2]` to `[cuda::]std::complex<T[2]>`, or casting `std::complex<T>` to `cuda::std::complex<T>` (this case can be handled by implicit conversion instead.

There's a big perf hit for doing the wrong thing here.

**Comment by gonzalobg** [comment]
@brycelelbach 

> I think we should break from what the standard requires here. The only case where it hits us is if someone is casting unaligned T[2] to [cuda::]std::complex<T[2]>, or casting std::complex<T> to cuda::std::complex<T>

Can you point to the paragraph of the standard that guarantees that this works?

**Comment by griwes** [comment]
The standard only requires reinterpreting complex as an array, not the other way around.

**Comment by leofang** [comment]
> My inclination is that we just make `cuda::std::complex` aligned to `sizeof(T)*2` and add implicit conversions from `std::complex` when being used with NVCC.

This is what Thrust currently does too, and it helps improve the performance.

**Comment by wmaxey** [comment]
This has been fixed in 1.6.0 with #172
